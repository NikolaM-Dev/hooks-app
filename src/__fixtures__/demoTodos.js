const demoTodos = [
  {
    id: 1,
    desc: 'Learn React',
    done: false,
  },
  {
    id: 2,
    desc: 'Learn Mongo',
    done: true,
  },
];

export default demoTodos;
