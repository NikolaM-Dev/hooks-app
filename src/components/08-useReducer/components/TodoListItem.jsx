import PropTypes from 'prop-types';

const TodoListItem = ({ todo, handleToggle, handleDelete, index }) => {
  return (
    <li key={todo.id} className="list-group-item">
      <p
        onClick={() => handleToggle(todo.id)}
        className={`${todo.done ? 'complete ' : ''}text-center`}
      >
        {index + 1}. {todo.desc}
      </p>
      <button onClick={() => handleDelete(todo.id)} className="btn btn-danger">
        Delete
      </button>
    </li>
  );
};

TodoListItem.propTypes = {
  todo: PropTypes.object.isRequired,
  handleToggle: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
  index: PropTypes.number.isRequired,
};

export default TodoListItem;
