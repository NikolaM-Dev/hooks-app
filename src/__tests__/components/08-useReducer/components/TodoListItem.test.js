import '@testing-library/jest-dom';
import { shallow } from 'enzyme';

import demoTodos from '__fixtures__/demoTodos';
import TodoListItem from 'components/08-useReducer/components/TodoListItem';

describe('Tests on <TodoListItem /> component', () => {
  const handleDelete = jest.fn();
  const handleToggle = jest.fn();
  const index = 0;
  const todo = demoTodos[0];

  let wrapper = shallow(
    <TodoListItem
      handleDelete={handleDelete}
      handleToggle={handleToggle}
      index={index}
      todo={todo}
    />
  );

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(
      <TodoListItem
        handleDelete={handleDelete}
        handleToggle={handleToggle}
        index={index}
        todo={todo}
      />
    );
  });

  it('should show <TodoListItem /> correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call the handleDelete function', () => {
    wrapper.find('button').simulate('click');

    expect(handleDelete).toHaveBeenCalledTimes(1);
    expect(handleDelete).toHaveBeenCalledWith(todo.id);
  });

  it('should call the handleToggle function', () => {
    wrapper.find('p').simulate('click');

    expect(handleToggle).toHaveBeenCalledTimes(1);
    expect(handleToggle).toHaveBeenCalledWith(todo.id);
  });

  it('should show the text correctly', () => {
    const text = wrapper.find('p').text().trim();
    const correctlyText = `${index + 1}. ${todo.desc}`;

    expect(text).toBe(correctlyText);
  });

  it('should have the complete class if todo.done is true', () => {
    const todo = demoTodos[1];

    const wrapper = shallow(
      <TodoListItem
        handleDelete={handleDelete}
        handleToggle={handleToggle}
        index={index}
        todo={todo}
      />
    );

    const haveCompleteClass = wrapper.find('p').hasClass('complete');

    expect(haveCompleteClass).toBeTruthy();
  });
});
