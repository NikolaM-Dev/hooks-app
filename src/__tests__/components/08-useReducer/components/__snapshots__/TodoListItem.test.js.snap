// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <TodoListItem /> component should show <TodoListItem /> correctly 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="text-center"
    onClick={[Function]}
  >
    1
    . 
    Learn React
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</li>
`;
